### PR TITLE
Set workdir at runtime and remove sed scripts from Dockerfiles and entry_point.sh.

### PR DESCRIPTION
**Pull request by nuclearsandwich**
These improvements are both extracted from #640 since the changes to the docker-run arguments are required for those changes but the Dockerfile and entry_point.sh improvements that they facilitate are not.

### CODE REVIEW

File: linux_docker_resources/entry_point.sh [hunk 1]
```diff
@@ -85,6 +85,4 @@ sed -i -e "s/rosbuild:x:$ORIG_GID:/rosbuild:x:$GID:/" /etc/group
 chown -R ${UID}:${GID} "${ORIG_HOME}"
 echo "done."
 
-cd "@workdir"
-
```
Comment (nuclearsandwich) [inline linux_docker_resources/entry_point.sh:89]: On reflection, removing this bit of the script actually makes bind-mounting it long-term a lot less important. I need to bind-mount it in #640 because I can't change the embedded script which is causing me problems with the workdir but we may not need to generally. I'm open to thoughts on pulling the bind-mount out of the permanent change entirely.

**Comment by nuclearsandwich** [comment]
The test_ci_linux job is presently occupied. I'd like to test the following parameter combos at least on on an "infra pass" which builds and test a minimal set of ROS 2, possibly just up to rcutils.

* Ubuntu Focal @ Rolling/master
* Ubuntu Jammy @ Rolling/master (sort of, planning to use #640 for this because of the snapshot work necessary while performing the 3.10 transition)
* RHEL 8 @ Rolling/master

There isn't any fundamental difference between Rolling and stable ROS distros for the changes here so I don't feel the need to run tests for Galactic or Foxy but I would if it assuages concerns.

File: linux_docker_resources/Dockerfile [hunk 2]
```diff
@@ -175,12 +175,6 @@ RUN useradd -u 1234 -m rosbuild
 RUN sudo -H -u rosbuild -- git config --global user.email "jenkins@ci.ros2.org"
 RUN sudo -H -u rosbuild -- git config --global user.name "Jenkins ROS 2"
 RUN echo 'rosbuild ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers
-WORKDIR "@workdir"
-
-# Add an entry point which changes rosbuild's UID from 1234 to the UID of the invoking user.
-# This means that the generated files will have the same ownership as the host OS user.
-ADD entry_point.sh /entry_point.sh
-RUN chmod 755 /entry_point.sh
```
[the remark this answers is not in the document]
Comment (cottsay) [reply linux_docker_resources/Dockerfile:183]: I wouldn't be opposed to writing out a placeholder bash script that no-ops, or even just symlinking something in the script's spot. It would be nice to be able to `docker run` this as a human being without all of the extra bind mounts for debugging purposes.

File: linux_docker_resources/entry_point.sh (hunk 1 above)
Comment (cottsay) [reply linux_docker_resources/entry_point.sh:89]: What purpose does this line serve, anyway? Doesn't the `WORKDIR` entry in the Dockerfile already ensure this is the case?

If we can really drop this, then I'd rather see the `ADD` kept in favor of more bind mounts.

**Review by cottsay — APPROVED** [review]
Some non-blocking thoughts, but this is overall goodness.

File: linux_docker_resources/entry_point.sh (hunk 1 above)
Comment (nuclearsandwich) [reply linux_docker_resources/entry_point.sh:89]: > What purpose does this line serve, anyway? 

It's redundant with `WORKDIR` (or `--workdir` in the new world order) and I'm pretty sure it was unnecessary when introduced in #248. I think the only reason to keep anything would a construct like

```
#!/bin/sh
PWD_BACKUP=$(pwd)
# script contents
cd $PWD_BACKUP
```
to make sure that a wayward script change doesn't change the directory when it shouldn't. However, I don't feel the need. I had this to say about it in ebf6e079ab10afddbba824ae1f34378ac459ea52 

> By eliminating this we remove a source of job-specific content in the
scripts and Dockerfile contents for CI.
> 
> By removing this we're also committing ourselves to being extra vigilant
that behavior inside the entry_point.sh script does not change the
directory context without restoring it. If there was concern about that
then an alternative would be to add

Comment (nuclearsandwich) [reply linux_docker_resources/entry_point.sh:89]: 16af6de3426ea2f5f2bcc4262990e5195c838669 effectively restores the original Dockerfile with one difference, added to a line comment on site.

File: linux_docker_resources/Dockerfile [hunk 3]
```diff
@@ -175,11 +175,10 @@ RUN useradd -u 1234 -m rosbuild
 RUN sudo -H -u rosbuild -- git config --global user.email "jenkins@ci.ros2.org"
 RUN sudo -H -u rosbuild -- git config --global user.name "Jenkins ROS 2"
 RUN echo 'rosbuild ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers
-WORKDIR "@workdir"
 
 # Add an entry point which changes rosbuild's UID from 1234 to the UID of the invoking user.
 # This means that the generated files will have the same ownership as the host OS user.
-ADD entry_point.sh /entry_point.sh
+COPY entry_point.sh /entry_point.sh
```
Comment (nuclearsandwich) [inline linux_docker_resources/Dockerfile:181]: The way we're using `ADD` wasn't that opaque but it hurt me to put an ADD back where a COPY would do.

[Docker's documentation](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#add-or-copy) recommends COPY over ADD unless you need ADD's specific features.

> Although ADD and COPY are functionally similar, generally speaking, COPY is preferred. That’s because it’s more transparent than ADD. COPY only supports the basic copying of local files into the container, while ADD has some features (like local-only tar extraction and remote URL support) that are not immediately obvious. Consequently, the best use for ADD is local tar file auto-extraction...

File: linux_docker_resources/Dockerfile-RHEL [hunk 4]
```diff
@@ -149,7 +148,7 @@ RUN chown rosbuild: /home/rosbuild/.colcon -R
 
 # Add an entry point which changes rosbuild's UID from 1234 to the UID of the invoking user.
 # This means that the generated files will have the same ownership as the host OS user.
-ADD entry_point.sh /entry_point.sh
+COPY entry_point.sh /entry_point.sh
```
Comment (nuclearsandwich) [inline linux_docker_resources/Dockerfile-RHEL:151]: This change is a drive-by improvement since I removed then re-added this line.
[Docker's documentation](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#add-or-copy) recommends COPY over ADD unless you need ADD's specific features.

>  Although ADD and COPY are functionally similar, generally speaking, COPY is preferred. That’s because it’s more transparent than ADD. COPY only supports the basic copying of local files into the container, while ADD has some features (like local-only tar extraction and remote URL support) that are not immediately obvious. Consequently, the best use for ADD is local tar file auto-extraction...

**Review by audrow — APPROVED** [review]
Looks good to me.
